src/common/bounded_key_counter.h

```diff
@@ -0,0 +1,187 @@
+ */
+template <typename Key, typename Count>
+class BoundedKeyCounter {
+  /// map type to associate keys with their counter values
+  using map_type = std::map<Key, Count>;
+  using value_type = typename map_type::value_type;
+
+  /// view type used for sorting key-value pairs by their counter value
+  using view_type = std::vector<const value_type*>;
+
+  /// maximum number of counters to store at once
+  const size_t bound;
+
+  /// map of counters, with a maximum size given by 'bound'
+  map_type counters;
+
+  /// storage for sorted key-value pairs
+  view_type sorted;
+
+  /// remembers how much of the range is actually sorted
+  typename view_type::iterator sorted_position;
+
+  /// invalidate view of sorted entries
+  void invalidate_sorted()
+  {
+    sorted_position = sorted.begin();
+    sorted.clear();
+  }
+
+  /// value_type comparison function for sorting in descending order
+  static bool value_greater(const value_type *lhs, const value_type *rhs)
+  {
+    return lhs->second > rhs->second;
+  }
+
+  /// map iterator that adapts value_type to value_type*
+  struct const_pointer_iterator : public map_type::const_iterator {
+    const_pointer_iterator(typename map_type::const_iterator i)
+      : map_type::const_iterator(i) {}
+    const value_type* operator*() const {
+      return &map_type::const_iterator::operator*();
+    }
+  };
+
+ protected:
+  /// return the number of sorted entries. marked protected for unit testing
+  size_t get_num_sorted() const
+  {
+    using const_iterator = typename view_type::const_iterator;
+    return std::distance<const_iterator>(sorted.begin(), sorted_position);
+  }
+
+ public:
+  BoundedKeyCounter(size_t bound)
+    : bound(bound)
+  {
+    sorted.reserve(bound);
+    sorted_position = sorted.begin();
+  }
+
+  /// return the number of keys stored
+  size_t size() const noexcept { return counters.size(); }
+
+  /// return the maximum number of keys
+  size_t capacity() const noexcept { return bound; }
+
+  /// increment a counter for the given key and return its value. if the key was
+  /// not present, insert it. if the map is full, return 0
+  Count insert(const Key& key, Count n = 1)
+  {
+    typename map_type::iterator i;
+
+    if (counters.size() < bound) {
+      // insert new entries at count=0
+      bool inserted;
+      std::tie(i, inserted) = counters.emplace(key, 0);
+      if (inserted) {
+        sorted.push_back(&*i);
+      }
+    } else {
+      // when full, refuse to insert new entries
+      i = counters.find(key);
+      if (i == counters.end()) {
+        return 0;
+      }
+    }
+
+    i->second += n; // add to the counter
+
+    // update sorted position if necessary. use a binary search for the last
+    // element in the sorted range that's greater than this counter
+    sorted_position = std::lower_bound(sorted.begin(), sorted_position,
+                                       &*i, &value_greater);
```

Comment:
@yehudasa this is the line i mentioned in the standup this morning. `sorted_position` is an iterator that tracks how much of the vector is sorted (i.e. all elements before the iterator)

as an example, assume that we have a vector `[a=5 b=4 c=3 #]` that is fully-sorted by descending value, with `#` to indicate that `sorted_position` is at the `end()`

if we insert a new key `d=4`, we call `sorted.push_back()` and the result is `[a=5 b=4 c=3 # d=4]`. this call to `lower_bound()` will move the `sorted_position` forward past all entries with `value <= 4` and end up with `[a=5 # b=4 c=3 d=4]`

if we instead update an existing key, like `b=7`, `lower_bound()` here would result in `[# a=5 b=7 c=3]`, a fully-unsorted vector

in either case, a later call to `get_highest()` with `count=2` would follow the `if (sorted_count < count) {` branch, which advances the `sorted_position` to index 2 and performs a `partial_sort()` to guarantee that the range `[0, 2)` is fully sorted

does that example cover the case that you were concerned about, or did i miss it?

the `GetNumSorted` test case covers both of these, using `get_num_sorted()` to infer the `sorted_position`

Reply:
@cbodley I don't remember what I had in mind, I think I missed the part where pos is updated after modifying an existing entry. So, iirc, the partial_sort would guarantee that the range is fully sorted, but you might have bigger entries outside of that range?

Reply:
the call to `std::partial_sort(sorted.begin(), sorted_position, sorted.end(), &value_greater);` passes iterators for the entire range, so it does guarantee that all elements up to `sorted_position` are `>=` than the elements that come after - it just doesn't guarantee that the elements after `sorted_position` are in the right order

Reply:
@cbodley what about a case where instead of updating existing key, we insert a new one with the same key?

Reply:
the keys are stored in the `counters` map, so this insert logic should prevent the `sorted` vector from containing duplicate keys:
```c++
      std::tie(i, inserted) = counters.emplace(key, 0);
      if (inserted) {
        sorted.push_back(&*i);
      }
```